### ISSUE TITLE
Circular dependencies in Python < 3.7 string annotations

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#710]
In GitLab by @crusaderky on Jul 13, 2019, 11:51

When a function of module A needs to use and return an object of module B, but module B already imports module A, you will typically see this pattern in Python >=3.5.2, <3.7:

```
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    # Circular dependency; module_b imports this module in its header
    from module_b import SomeClass

def f() -> "SomeClass":
    from module_b import SomeClass
    return SomeClass(1, 2, 3)
```

flake8 3.7.7 produces an error when parsing this file:
```
module_a.py:8:5: F811 redefinition of unused 'SomeClass' from line 5
```

Notably, flake8 does not complain when the same is rewritten using delayed annotation, which however requires Python 3.7:

```
from __future__ import annotations
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    # Circular dependency; module_b imports this module in its header
    from module_b import SomeClass

def f() -> SomeClass:
    from module_b import SomeClass
    return SomeClass(1, 2, 3)
```

**Comment by asottile** [comment]
In GitLab by @crusaderky on Jul 13, 2019, 11:51

changed title from **Circular dependencies in Python < 3.7 {-delayed-} annotations** to **Circular dependencies in Python < 3.7 {+string+} annotations**

**Comment by asottile** [comment]
In GitLab by @crusaderky on Jul 13, 2019, 11:52

changed the description

**Comment by asottile** [comment]
In GitLab by @asottile on Jul 13, 2019, 11:55

flake8 is just a framework that combines other linters / formatters

you want `pyflakes` which implements the `F*` lints

**Comment by asottile** [comment]
In GitLab by @asottile on Jul 13, 2019, 11:55

closed

**Comment by asottile** [comment]
In GitLab by @crusaderky on Jul 13, 2019, 13:21

https://github.com/PyCQA/pyflakes/issues/453
